src/coreclr/src/debug/debug-pal/win/diagnosticsipc.cpp

```diff
@@ -87,45 +116,163 @@ IpcStream *IpcStream::DiagnosticsIpc::Accept(ErrorCallback callback) const
                 if (callback != nullptr)
                     callback("A client process failed to connect.", errorCode);
                 ::CloseHandle(hPipe);
+                delete pStream;
                 return nullptr;
         }
     }
 
-    return new IpcStream(hPipe);
+    return pStream;
+}
+
+IpcStream *IpcStream::DiagnosticsIpc::Connect(ErrorCallback callback)
+{
+    _ASSERTE(mode == ConnectionMode::CLIENT);
+    if (mode != ConnectionMode::CLIENT)
+    {
+        if (callback != nullptr)
+            callback("Cannot call connect on a client connection", 0);
+        return nullptr;
+    }
+
+    HANDLE hPipe = ::CreateFileA( 
+        _pNamedPipeName,                    // pipe name 
+        PIPE_ACCESS_DUPLEX,                 // pipe access
+        0,                                  // no sharing 
+        NULL,                               // default security attributes
+        OPEN_EXISTING,                      // opens existing pipe 
+        0,                                  // default attributes 
+        NULL);                              // no template file
+
+    if (hPipe == INVALID_HANDLE_VALUE)
+    {
+        if (callback != nullptr)
+            callback("Failed to connect to named pipe.", ::GetLastError());
+        return nullptr;
+    }
+
+    return new IpcStream(hPipe, mode);
 }
 
 void IpcStream::DiagnosticsIpc::Close(ErrorCallback)
 {
 }
 
+IpcStream::IpcStream(HANDLE hPipe, DiagnosticsIpc::ConnectionMode mode) :
+    _hPipe(hPipe), 
+    _mode(mode) 
+{
+    if (_mode == DiagnosticsIpc::ConnectionMode::SERVER)
+        _oOverlap.hEvent = CreateEvent(NULL, true, false, NULL);
```

Comment:
This handle is leaking when IpcStream is destroyed

Reply:
Good catch, thanks!